sqlx-core/src/postgres/protocol/sasl.rs

```diff
@@ -0,0 +1,205 @@
+use crate::io::BufMut;
+use crate::postgres::connection::PgConnection;
+use crate::postgres::protocol::authentication::Authentication::SaslContinue;
+use crate::postgres::protocol::Encode;
+use crate::postgres::protocol::Message;
+use crate::Result;
+use byteorder::NetworkEndian;
+use hmac::{Hmac, Mac};
+use rand::Rng;
+use sha2::{Digest, Sha256};
+
+pub struct SaslInitialResponse {
+    // pub username: String,
+    // pub passord: String,
+    pub s: String,
+}
+
+impl Encode for SaslInitialResponse {
+    fn encode(&self, buf: &mut Vec<u8>) {
+        buf.push(b'p');
+        buf.put_u32::<NetworkEndian>(4u32 + self.s.as_str().as_bytes().len() as u32 + 14u32 + 4u32);
+        buf.put_str_nul("SCRAM-SHA-256");
+        buf.put_u32::<NetworkEndian>(self.s.as_str().as_bytes().len() as u32);
+        buf.extend_from_slice(self.s.as_str().as_bytes());
+    }
+}
+
+pub struct SaslResponse {
+    pub s: String,
+}
+
+impl Encode for SaslResponse {
+    fn encode(&self, buf: &mut Vec<u8>) {
+        buf.push(b'p');
+        buf.put_u32::<NetworkEndian>(4u32 + self.s.as_str().as_bytes().len() as u32);
+        buf.extend_from_slice(self.s.as_str().as_bytes());
+    }
+}
+
+static GS2_HEADER: &'static str = "n,,";
+static CHANNEL_ATTR: &'static str = "c";
+static USERNAME_ATTR: &'static str = "n";
+static CLIENT_PROOF_ATTR: &'static str = "p";
+static NONCE_ATTR: &'static str = "r";
+
+pub fn nonce() -> String {
+    let mut rng = rand::thread_rng();
+    let count = rng.gen_range(64, 128);
+    // printable = %x21-2B / %x2D-7E
+    // ;; Printable ASCII except ",".
+    // ;; Note that any "printable" is also
+    // ;; a valid "value".
+    let nonce: String = std::iter::repeat(())
+        .map(|()| {
+            let mut c = rng.gen_range(0x21, 0x7F) as u8;
+
+            while c == 0x2C {
+                c = rng.gen_range(0x21, 0x7F) as u8;
+            }
+
+            c
+        })
+        .take(count)
+        .map(|c| c as char)
+        .collect();
+
+    rng.gen_range(32, 128);
+    format!("{}={}", NONCE_ATTR, nonce)
+}
+
+pub async fn sasl_auth<T: AsRef<str>>(
+    conn: &mut PgConnection,
```

Comment:
This function should be in or around `connection`. You don't want the protocol module to depend on the connection types.